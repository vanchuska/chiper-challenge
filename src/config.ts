import { registerAs } from '@nestjs/config';

export default registerAs('config', () => {
  return {
    food_truck_api_url: process.env.FOOD_TRUCK_API_URL,
    food_truck_app_token: process.env.FOOD_TRUCK_APP_TOKEN,

  
  };
});
