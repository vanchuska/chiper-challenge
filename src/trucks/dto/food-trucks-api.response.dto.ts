export interface FoodTruckResponse {
    type:     string;
    features: Feature[];
    crs:      CRS;
}

 interface CRS {
    type:       string;
    properties: CRSProperties;
}

 interface CRSProperties {
    name: string;
}

 interface Feature {
    type:       string;
    geometry:   Geometry;
    properties: FeatureProperties;
}

 interface Geometry {
    type:        string;
    coordinates: number[];
}

 interface FeatureProperties {
    location_state:      string;
    x:                   string;
    location_zip:        string;
    applicant:           string;
    locationdescription: string;
    dayshours:           string;
    latitude:            string;
    y:                   string;
    blocklot:            string;
    location_address:    string;
    noisent:             Date;
    location_city:       string;
    cnn:                 string;
    objectid:            string;
    longitude:           string;
    block:               string;
    permit:              string;
    status:              string;
    facilitytype:        string;
    schedule:            string;
    lot:                 string;
    address:             string;
    approved:            Date;
    fooditems:           string;
    received:            string;
    expirationdate:      Date;
    priorpermit:         string;
}