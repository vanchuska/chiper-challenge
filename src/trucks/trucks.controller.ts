import { Controller, Get, Logger, Res,HttpStatus } from '@nestjs/common';
import { response } from 'express';
import { TrucksService } from './trucks.service';

@Controller('trucks')
export class TrucksController {
  private readonly logger = new Logger('trucks');
  constructor(private readonly trucksService: TrucksService) {}

  @Get()
  getFoodTrucks() {
    this.logger.warn('getFoodTrucks');
    return this.trucksService.getFoodTrucks();
  }

  // @Get('api')
  // getTrucks(@Res() response) {
  //   this.logger.warn('getFoodTrucks');
  //   return this.trucksService.getFoodTrucks().then(res => {
  //     response.status(HttpStatus.CREATED).json(res);
  //   }).catch((err)=> {
  //     response.status(HttpStatus.NO_CONTENT).json(err);
  //   })
     
  // }
}
