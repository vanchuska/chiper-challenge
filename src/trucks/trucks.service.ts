import { ForbiddenException, Inject, Injectable } from '@nestjs/common';
import { AxiosResponse } from 'axios';
import { HttpService } from '@nestjs/axios';
import { map, catchError, lastValueFrom, firstValueFrom } from 'rxjs';
import { ConfigType } from '@nestjs/config';
import config from '../config';
import { FoodTruckResponse } from './dto/food-trucks-api.response.dto';

@Injectable()
export class TrucksService {
  constructor(
    private readonly httpService: HttpService,
    @Inject(config.KEY) private configService: ConfigType<typeof config>,
  ) {}

  async getTrucks() {
    const request = this.httpService
      .get(
        this.configService.food_truck_api_url +
          '?$$app_token=' +
          this.configService.food_truck_app_token,
      )
      .pipe(map((res) => res.data))
      .pipe(
        catchError(() => {
          throw new ForbiddenException('API not available');
        }),
      );

    const trucks = await lastValueFrom(request);

    return {
      data: {
        trucks,
      },
    };
  }

  async getFoodTrucks(): Promise<FoodTruckResponse> {
    let resp: FoodTruckResponse;
    const result = await firstValueFrom(
      this.httpService.get<FoodTruckResponse>(
        this.configService.food_truck_api_url +
          '?$$app_token=' +
          this.configService.food_truck_app_token,
      ),
    );
    const { data, status } = result;
    if (status === 200) {
      resp = data;
    }
    return resp;
  }
}
