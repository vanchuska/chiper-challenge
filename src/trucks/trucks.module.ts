import { Module } from '@nestjs/common';
import { TrucksService } from './trucks.service';
import { TrucksController } from './trucks.controller';
import { HttpModule } from '@nestjs/axios';
import { HttpConfiService } from 'src/httpService.config';

@Module({
  imports: [
    HttpModule.registerAsync({
      useClass :HttpConfiService,
    }),
    TrucksModule,
    
  ],
  controllers: [TrucksController],
  providers: [TrucksService]
})
export class TrucksModule {}
