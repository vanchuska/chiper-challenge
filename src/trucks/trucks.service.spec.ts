import { HttpService } from '@nestjs/axios';
import { ConfigType } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import { error } from 'console';
import { of, throwError } from 'rxjs';
import config from '../config';
import { FoodTruckResponse } from './dto/food-trucks-api.response.dto';
import { TrucksService } from './trucks.service';

describe('TrucksService', () => {
  let trucksService: TrucksService;
  let httpService: HttpService;
  let configService: ConfigType<typeof config>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        TrucksService,
        {
          provide: HttpService,
          useValue: {
            get: jest.fn(),
          },
        },
        {
          provide: config.KEY,
          useValue: {},
        },
      ],
    }).compile();

    trucksService = module.get<TrucksService>(TrucksService);
    httpService = module.get<HttpService>(HttpService);
    configService = module.get(config.KEY);
  });

  it('should be defined', () => {
    expect(trucksService).toBeDefined();
    expect(httpService).toBeDefined();
    expect(configService).toBeDefined();
  });

  describe('getFoodTrucks', () => {
    it('should return a FoodTruck Response', async () => {
      // Arrange
      const expectedData: FoodTruckResponse = {
        type: 'FeatureCollection',
        features: [
          {
            type: 'Feature',
            geometry: {
              type: 'Point',
              coordinates: [-122.39581105302317, 37.794331003246846],
            },
            properties: {
              location_state: '',
              x: '6013916.72',
              location_zip: '',
              applicant: 'Ziaurehman Amini',
              locationdescription: 'MARKET ST: DRUMM ST intersection',
              dayshours: null,
              latitude: '37.794331003246846',
              y: '2117244.027',
              blocklot: '0234017',
              location_address: '',
              noisent: null,
              location_city: '',
              cnn: '30727000',
              objectid: '735318',
              longitude: '-122.39581105302317',
              block: '0234',
              permit: '15MFF-0159',
              status: 'REQUESTED',
              facilitytype: 'Push Cart',
              schedule:
                'http://bsm.sfdpw.org/PermitsTracker/reports/report.aspx?title=schedule&report=rptSchedule&params=permit=15MFF-0159&ExportPDF=1&Filename=15MFF-0159_schedule.pdf',
              lot: '017',
              address: '5 THE EMBARCADERO',
              approved: null,
              fooditems: null,
              received: '20151231',
              expirationdate: new Date('2016-03-15T00:00:00.000'),
              priorpermit: '0',
            },
          },
        ],
        crs: {
          type: 'name',
          properties: {
            name: 'urn:ogc:def:crs:OGC:1.3:CRS84',
          },
        },
      };

      jest.spyOn(httpService, 'get').mockReturnValueOnce(
        of({
          data: expectedData,
          status: 200,
          statusText: 'OK',
          config: {},
          headers: {},
        }),
      );
      // Act
      const result = await trucksService.getFoodTrucks();
      // Assert
      expect(result).toEqual(expectedData);
      expect(result).toMatchSnapshot();
    });

    it('should return a empty Response', async () => {
      // Arrange
      const expectedData: FoodTruckResponse = undefined;

      jest.spyOn(httpService, 'get').mockReturnValueOnce(
        of({
          data: expectedData,
          status: 204,
          statusText: 'NO CONTENT',
          config: {},
          headers: {},
        }),
      );
      // Act
      const result = await trucksService.getFoodTrucks();
      // Assert
      expect(result).toEqual(expectedData);
    });

    it('should throw an unexpected error', () => {
      // Arrange
      jest
        .spyOn(httpService, 'get')
        .mockReturnValueOnce(throwError(() => new Error('Unexpected Error')));
      // Assert
      expect(trucksService.getFoodTrucks()).rejects.toThrowError(
        'Unexpected Error',
      );
    });
  });
});
