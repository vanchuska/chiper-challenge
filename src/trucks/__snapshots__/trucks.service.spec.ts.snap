// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrucksService getFoodTrucks should return a FoodTruck Response 1`] = `
Object {
  "crs": Object {
    "properties": Object {
      "name": "urn:ogc:def:crs:OGC:1.3:CRS84",
    },
    "type": "name",
  },
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          -122.39581105302317,
          37.794331003246846,
        ],
        "type": "Point",
      },
      "properties": Object {
        "address": "5 THE EMBARCADERO",
        "applicant": "Ziaurehman Amini",
        "approved": null,
        "block": "0234",
        "blocklot": "0234017",
        "cnn": "30727000",
        "dayshours": null,
        "expirationdate": 2016-03-15T05:00:00.000Z,
        "facilitytype": "Push Cart",
        "fooditems": null,
        "latitude": "37.794331003246846",
        "location_address": "",
        "location_city": "",
        "location_state": "",
        "location_zip": "",
        "locationdescription": "MARKET ST: DRUMM ST intersection",
        "longitude": "-122.39581105302317",
        "lot": "017",
        "noisent": null,
        "objectid": "735318",
        "permit": "15MFF-0159",
        "priorpermit": "0",
        "received": "20151231",
        "schedule": "http://bsm.sfdpw.org/PermitsTracker/reports/report.aspx?title=schedule&report=rptSchedule&params=permit=15MFF-0159&ExportPDF=1&Filename=15MFF-0159_schedule.pdf",
        "status": "REQUESTED",
        "x": "6013916.72",
        "y": "2117244.027",
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;
